plutus-contract/src/Plutus/Trace/Types.hs

```diff
@@ -0,0 +1,269 @@
+      Seq.EmptyL  -> NoMoreThreads
+
+-- Version 2
+-- Reified events
+
+data ThreadId = Networking | UserInput | OtherThread Int
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (Hashable)
+
+data ThreadMsg = EndpointCall String String | OwnPkResponse PubKey
+
+data InboxMsg = InboxMsg {imEnqueuedAt :: Slot, imMessage :: ThreadMsg}
+
+data OutboxMsg = OutboxMsg {oRecipient :: ThreadId, oMsg :: ThreadMsg}
+
+newtype Worker effs = Worker {unWorker :: InboxMsg -> Eff effs [OutboxMsg]}
+
+data MBState effs
+  = MBState
+      { mbWorkers :: HashMap ThreadId (Worker effs),
+        mbInboxes :: HashMap ThreadId (Seq InboxMsg)
+      }
+
+makeLensesFor [("mbWorkers", "workers"), ("mbInboxes", "inboxes")] ''MBState
+
+data MBError =
+    NoMessagesForThread ThreadId
+    | NoWorkerForThread ThreadId
+
+-- | Take the first message from the thread's inbox.
+dequeueMsg :: forall effs. ThreadId -> Eff (State (MBState effs) ': Error MBError ': effs) InboxMsg
+dequeueMsg tid = do
+  (msg, rest) <- do
+    r <- gets @(MBState effs) (view (inboxes . at tid)) >>= maybe (throwError $ NoMessagesForThread tid) pure
+    case Seq.viewl r of
+        Seq.EmptyL    -> throwError $ NoMessagesForThread tid
+        a Seq.:< rest -> pure (a, rest)
+  if (Seq.null rest)
+    then modify @(MBState effs) (inboxes . at tid .~ Just rest)
+    else modify @(MBState effs) (inboxes . at tid .~ Nothing)
+  pure msg
+
+-- | Route an 'OutboxMessage' to the inbox of its recipient
+enqueueMsg :: forall effs. OutboxMsg -> Eff (State (MBState effs) ': Error MBError ': effs) ()
+enqueueMsg OutboxMsg{oRecipient, oMsg} = modify @(MBState effs) (over (inboxes . at oRecipient) addToInbox) where
+    inboxMessage = InboxMsg{imEnqueuedAt = 0, imMessage = oMsg}
+
+    addToInbox :: Maybe (Seq InboxMsg) -> Maybe (Seq InboxMsg)
+    addToInbox Nothing  = Just $ Seq.singleton inboxMessage
+    addToInbox (Just s) = Just $ s Seq.|> inboxMessage
+
+-- | Get the 'Worker effs' for a thread ID.
+getThread :: forall effs. ThreadId -> Eff (State (MBState effs) ': Error MBError ': effs) (Worker effs)
+getThread tid = do
+    r <- gets @(MBState effs) (view $ workers . at tid)
+    case r of
+        Nothing -> throwError $ NoWorkerForThread tid
+        Just w  -> pure w
+
+-- | Take the first message for this worker, process it, and add the resulting
+--   messages to the appropriate queues
+handleWorker :: ThreadId -> Eff (State (MBState effs) ': Error MBError ': effs) ()
+handleWorker tid = do
+  msg <- dequeueMsg tid
+  Worker thread <- getThread tid
+  newMSGs <- raise $ raise $ thread msg
+  traverse_ enqueueMsg newMSGs
+  pure ()
+
+-- | Process the first message in each inbox, until all inboxes are empty.
+loop2 :: forall effs. Eff (State (MBState effs) ': Error MBError ': effs) ()
+loop2 = do
+  currentState <- get @(MBState effs)
+  unless (HashMap.null $ mbInboxes currentState) $ do
+    State.gets @(MBState effs) (HashMap.keysSet . view inboxes) >>= traverse_ handleWorker
```

Comment:
`handleWorker` will throw when it has no messages, rather than skipping, so I don't think this will work right.

Reply:
The map only contains non-empty sequences (when a worker has no more messages then the entry is deleted from the map)

This is so that we can easily check if all the threads are done, via `HashMap.null`

Reply:
I should probably look for non-empty sequence/queue implementation

Reply:
Ah right, you only traverse over the threads with non-empty messageboxes. Could be simpler, but not a big deal.